config/config.go

```diff
@@ -225,6 +207,234 @@ func validateReferencesExist(c configpb.Configuration) error {
+				},
+			)
+		}
+
+	}
+
+	// test_name_config should have a matching number of format strings and name elements.
+	if tg.TestNameConfig != nil {
+		nameFormat := tg.GetTestNameConfig().GetNameFormat()
+		nameElements := tg.GetTestNameConfig().GetNameElements()
+
+		if len(nameElements) == 0 {
+			mErr = multierror.Append(mErr, &ConfigError{tg.Name, "TestGroup", "TestNameConfig.NameElements must be specified."})
+		}
+
+		if nameFormat == "" {
+			mErr = multierror.Append(mErr, &ConfigError{tg.Name, "TestGroup", "TestNameConfig.NameFormat must be specified."})
+		} else {
+			if got, want := len(nameElements), strings.Count(nameFormat, "%"); got != want {
+				mErr = multierror.Append(
+					mErr,
+					&ConfigError{tg.Name, "TestGroup", fmt.Sprintf("TestNameConfig has %d elements, format %s wants %d", got, nameFormat, want)},
+				)
+			}
+			elements := make([]interface{}, 0)
+			for range nameElements {
+				elements = append(elements, "")
+			}
+			s := fmt.Sprintf(nameFormat, elements...)
+			if strings.Contains(s, "%!") {
+				return &ConfigError{tg.Name, "TestGroup", fmt.Sprintf("Number of format strings and name_elements must match; got %s (%d).", s, len(elements))}
+			}
+		}
+	}
+
+	return mErr.ErrorOrNil()
+}
+
+func validateDashboardTab(dt *configpb.DashboardTab) error {
+	mErr := &multierror.Error{}
+
+	// Check that required fields are a non-zero-value.
+	if dt.TestGroupName == "" {
+		mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", "test_group_name can't be empty"})
+	}
+
+	// A Dashboard Tab can't be named the same as the default 'Summary' tab.
+	if dt.Name == "Summary" {
+		mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", "tab can't be named 'Summary'"})
+	}
+
+	// TabularNamesRegex should be valid and have capture groups defined.
+	if dt.TabularNamesRegex != "" {
+		regex, err := regexp.Compile(dt.TabularNamesRegex)
+		if err != nil {
+			mErr = multierror.Append(
+				mErr,
+				&ConfigError{dt.Name, "DashboardTab", fmt.Sprintf("Invalid regex %s: %v", dt.TabularNamesRegex, err)})
+		} else {
+			if regex.NumSubexp() != len(regex.SubexpNames()) {
+				mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", "all tabular_name_regex capture groups must be named"})
+			}
+			if len(regex.SubexpNames()) < 1 {
+				mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", "tabular_name_regex requires at least one capture group"})
+			}
+		}
+	}
+
+	// Email address for alerts should be valid.
+	if dt.GetAlertOptions().GetAlertMailToAddresses() != "" {
+		err := validateEmails(dt.GetAlertOptions().GetAlertMailToAddresses())
+		if err != nil {
+			mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", err.Error()})
+		}
+	}
+
+	return mErr.ErrorOrNil()
+}
+
+func validateEntityConfigs(c configpb.Configuration) error {
+	mErr := &multierror.Error{}
```

Comment:
`var mErr error`

https://github.com/hashicorp/go-multierror#usage

Reply:
Has to be a multierror error for ErrorOrNil, but changed.

Reply:
I don't see why all this code can't just be the more idiomatic:
```golang
var merr error
if err := foo(); err != nil {
  merr = multierror.Append(merr, err)
}
if err := bar(); err != nil {
  merr = multierror.Append(merr, err)
}
// etc
return merr
```

Reply:
Ah, misunderstood. Done!